src/persist-client/src/impl/compact.rs

```diff
@@ -0,0 +1,294 @@
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use mz_persist::indexed::columnar::ColumnarRecordsVecBuilder;
+use mz_persist::location::Blob;
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::Timestamp;
+use tokio::runtime::Handle;
+use tracing::{debug_span, warn, Instrument, Span};
+
+use crate::batch::BatchParts;
+use crate::r#impl::machine::Machine;
+use crate::r#impl::state::HollowBatch;
+use crate::r#impl::trace::FueledMergeRes;
+use crate::read::fetch_batch_part;
+use crate::{Metrics, PersistConfig, ShardId};
+
+/// A request for compaction.
+///
+/// This is similar to FueledMergeReq, but intentionally a different type. If we
+/// move compaction to an rpc server, this one will become a protobuf; the type
+/// parameters will become names of codecs to look up in some registry.
+#[derive(Debug, Clone)]
+pub struct CompactReq<T> {
+    /// The shard the input and output batches belong to.
+    pub shard_id: ShardId,
+    /// A description for the output batch.
+    pub desc: Description<T>,
+    /// The updates to include in the output batch. Any data in these outside of
+    /// the output descriptions bounds should be ignored.
+    pub inputs: Vec<HollowBatch<T>>,
+}
+
+/// A response from compaction.
+#[derive(Debug)]
+pub struct CompactRes<T> {
+    /// The compacted batch.
+    pub output: HollowBatch<T>,
+}
+
+/// A service for performing physical and logical compaction.
+///
+/// This will possibly be called over RPC in the future. Physical compaction is
+/// merging adjacent batches. Logical compaction is advancing timestamps to a
+/// new since and consolidating the resulting updates.
+#[derive(Debug, Clone)]
+pub struct Compactor {
+    cfg: PersistConfig,
+    blob: Arc<dyn Blob + Send + Sync>,
+    metrics: Arc<Metrics>,
+}
+
+impl Compactor {
+    pub fn new(
+        cfg: PersistConfig,
+        blob: Arc<dyn Blob + Send + Sync>,
+        metrics: Arc<Metrics>,
+    ) -> Self {
+        Compactor { cfg, blob, metrics }
+    }
+
+    pub fn compact_and_apply<K, V, T, D>(&self, machine: &Machine<K, V, T, D>, req: CompactReq<T>)
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        assert_eq!(req.shard_id, machine.shard_id());
+
+        // Run some initial heuristics to ignore some requests for compaction.
+        // We don't gain much from e.g. compacting two very small batches that
+        // were just written, but it does result in non-trivial blob traffic
+        // (especially in aggregate). This heuristic is something we'll need to
+        // tune over time.
+        #[allow(unused_assignments)]
+        let mut should_compact =
+            req.inputs.len() >= 16 || req.inputs.iter().map(|x| x.len).sum::<usize>() >= 128;
+        // TODO: Enable the heuristics after we've gotten a sense for where
+        // we're starting.
```

Comment:
my convention (just so you know for the future) is that WIP generally means I intended to do it before merging and TODO means not. though, funny enough, I've started to warm up to the idea of enabling them from the beginning. I don't see any real reason not to and it might help avoid giving a bad initial impression of persist compaction to the rest of the company

metrics are a good idea! done!